File: src/Binary/diameterOFBinaryTree.java
package Binary;

import BinaryTree.NodeB;

public class diameterOFBinaryTree {


    public static int diameter(NodeB root) {
        if (root == null) {
            return -1;
        }
        int leftHeight = height(root.left);
        int rightHeight = height(root.right);
        int leftDiameter = diameter(root.left);
        int rightDiameter = diameter(root.right);
        int sp = leftHeight + rightHeight + 2;
        int maxDiameter = Math.max(sp, Math.max(leftDiameter, rightDiameter));
        return maxDiameter;
    }

    public static int height(NodeB root) {
        if (root == null) {
            return -1;
        }
        int leftHeight = height(root.left);
        int rightHeight = height(root.right);
        return Math.max(leftHeight, rightHeight) + 1;
    }

    public static void main(String[] args) {
        NodeB root = new NodeB(1);
        root.left = new NodeB(2);
        root.right = new NodeB(3);
        root.left.left = new NodeB(4);

        System.out.println(diameter(root));
    }
}

File: src/BinaryTree/diameterOfbinaryTree.cpp
#include <bits/stdc++.h>
using namespace std;

struct node
{
    int data;
    node *left, *right;
    node(int x)
    {
        data = x;
        left = right = NULL;
    }
};
int height(node *root)
{
    if (root == NULL)
        return 0;
    int left = height(root->left);
    int right = height(root->right);
    return max(left, right) + 1;
}

int diameter(node *root)
{
    if (root == NULL)
        return 0;
    int d1 = 1 + height(root->left) + height(root->right);
    int d2 = diameter(root->left);
    int d3 = diameter(root->right);
    return max(d1, max(d2, d3));
}
int res = 0;
int height2(node *root)
{
    if (root == NULL)
        return 0;
    int lh = height2(root->left);
    int rh = height2(root->right);
    res = max(res, 1 + lh + rh);
    cout<<res<<" ";
    
    return 1 + max(lh, rh);
   
}
int main()
{
    node *root = new node(1);
    root->left = new node(2);
    root->right = new node(3);
    root->left->left = new node(4);
    root->left->right = new node(5);
    root->right->left = new node(6);

    cout << height2(root);
    return 0;
}